facilitator/src/transport/gcs.rs

```diff
@@ -0,0 +1,601 @@
+        let upload_session_uri = match http_response.header("Location") {
+            Some(location) => location,
+            None => {
+                return Err(anyhow!(
+                    "no Location header in response when initiating streaming transfer"
+                ))
+            }
+        };
+
+        Ok(StreamingTransferWriter {
+            minimum_upload_chunk_size,
+            buffer: Vec::with_capacity(minimum_upload_chunk_size * 2),
+            object_upload_position: 0,
+            upload_session_uri: upload_session_uri.to_owned(),
+        })
+    }
+
+    fn upload_chunk(&mut self, last_chunk: bool) -> Result<()> {
+        if self.buffer.is_empty() {
+            return Ok(());
+        }
+
+        let chunk = mem::replace(
+            &mut self.buffer,
+            Vec::with_capacity(self.minimum_upload_chunk_size * 2),
+        );
+        let chunk_len = chunk.len();
+        // We may not yet be at the last chunk, and hence may not know the total
+        // length of eventual complete object, but this variable the total
+        // length of all the bytes fed into write(), even if we haven't uploaded
+        // them all yet.
+        let object_length_thus_far = self.object_upload_position + chunk_len;
+
+        // When this is the last piece being uploaded, the Content-Range header
+        // should include the total object size, but otherwise should have * to
+        // indicate to GCS that there is an unknown further amount to come.
+        // https://cloud.google.com/storage/docs/streaming#streaming_uploads
+        let content_range_header_total_length_field = if last_chunk {
+            format!("{}", object_length_thus_far)
+        } else {
+            "*".to_owned()
+        };
+        let content_range = format!(
+            "bytes {}-{}/{}",
+            self.object_upload_position,
+            object_length_thus_far - 1,
+            content_range_header_total_length_field
+        );
+
+        // ureq::Request::send takes possession of the argument so we must clone
+        // here, in case the entire chunk doesn't get uploaded successfully.
+        let body = Cursor::new(chunk.clone());
+
+        let http_response = ureq::put(&self.upload_session_uri)
+            .set("Content-Length", &format!("{}", chunk_len))
+            .set("Content-Range", &content_range)
+            // By default, ureq will wait forever to connect or read
+            .timeout_connect(10_000) // ten seconds
+            .timeout_read(10_000) // ten seconds
+            .send(body);
+
+        // On success we expect HTTP 308 Resume Incomplete and a Range: header,
+        // unless this is the last part and the server accepts the entire
+        // provided Content-Range, in which case it's HTTP 200, or 201 (?).
+        // https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload
+        match http_response.status() {
+            200 | 201 => {
+                if !last_chunk {
```

Comment:
In a big `match` like this, having a fallthrough makes it hard to reason about. I'd make the `match` the final item in the function, and split this into two clauses:

```
200 | 201 if last_chunk => Ok(()),
200 | 201 if !last_chunk => Err(...chunks remaining...),
```